source/core.py

```diff
@@ -432,27 +432,37 @@ def handlePowerStatusChange(self):
 	log.debug("Initializing core pump")
 	class CorePump(wx.Timer):
 		"Checks the queues and executes functions."
+		_inNotify=False
 		def Notify(self):
-			global _isPumpPending
-			_isPumpPending = False
-			watchdog.alive()
+			if self._inNotify:
+				# #7104: Timers in wxPython 4 are now re-entrant. However, wx.Yield is not allowed to be called recursively.
+				# Therefore, stop this timer callback from ever being re-entered.
+				# Note that If code does try to request a pump from within this call, it will be detected and re-queued by restarting the timer.
+				return
 			try:
-				if touchHandler.handler:
-					touchHandler.handler.pump()
-				JABHandler.pumpAll()
-				IAccessibleHandler.pumpAll()
-				queueHandler.pumpAll()
-				mouseHandler.pumpAll()
-				braille.pumpAll()
-			except:
-				log.exception("errors in this core pump cycle")
-			baseObject.AutoPropertyObject.invalidateCaches()
-			watchdog.asleep()
-			if _isPumpPending and not _pump.IsRunning():
-				# #3803: A pump was requested, but the timer was ignored by a modal loop
-				# because timers aren't re-entrant.
-				# Therefore, schedule another pump.
-				_pump.Start(PUMP_MAX_DELAY, True)
+				self._inNotify=True
+				global _isPumpPending
+				_isPumpPending = False
+				watchdog.alive()
+				try:
+					if touchHandler.handler:
+						touchHandler.handler.pump()
+					JABHandler.pumpAll()
+					IAccessibleHandler.pumpAll()
+					queueHandler.pumpAll()
+					mouseHandler.pumpAll()
+					braille.pumpAll()
+				except:
+					log.exception("errors in this core pump cycle")
+				baseObject.AutoPropertyObject.invalidateCaches()
+				watchdog.asleep()
+				if _isPumpPending and not _pump.IsRunning():
+					# #3803: A pump was requested, but the timer was ignored by a modal loop
+					# because timers aren't re-entrant.
+					# Therefore, schedule another pump.
+					_pump.Start(PUMP_MAX_DELAY, True)

```

Comment:
Shouldn't this logic / comment change now that wx.Timer is reentrant? I can't find any documentation about the recursive, reentrant, or thread safety of wx.Timer. Are there references for this somewhere. I would like to confirm I understand when used in the context of python and wx.

Reply:
That logic is there to catch situations where code that is run within the core pump requests a pump. As the timer is not re-entrant (either by design, or by our new code just added), the timer needs to be restarted to cause another pump after this one completes.
I'm not sure if there is official documentation on why wx.Timer is now not re-entrant, but see comments from @jcsteh in #7077 where he talks about the fact this has changed and an email thread on the wxPython email list I think.

Reply:
Ok, I have re-read through those comments. I think I understand the problem. My initial concern remains though. In the comment on line 461 we state that timers are not re-entrant, in the comment on line 438 we say they are. It seems the comment on 461 needs to be adjusted.